File: src/main/java/net/cakemc/database/file/NioFile.java
package net.cakemc.database.file;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.LinkOption;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;

/**
 * The type Nio file.
 */
public class NioFile extends AbstractDatabaseFile {

    /**
     * The Path.
     */
    protected final Path path;
    /**
     * The Default memory file.
     */
    protected MemoryFile defaultMemoryFile;
    /**
     * The Compressed.
     */
    protected final boolean compressed;

    /**
     * Instantiates a new Nio file.
     *
     * @param path       the path
     * @param compressed the compressed
     */
    public NioFile(Path path, boolean compressed) {
        this.path = path;
        this.compressed = compressed;
    }

    @Override
    public void read() throws IOException {
        this.defaultMemoryFile = new DefaultMemoryFile(path, Files.readAllBytes(path), compressed);
    }

    @Override
    public void write() throws Throwable {
         byte[] data = (this.defaultMemoryFile == null ?
                 defaultMemoryFile = new DefaultMemoryFile(path, new byte[0], compressed) :
                 defaultMemoryFile)
                 .getData();

        if (!exists()) {
            Files.write(path, data, StandardOpenOption.CREATE_NEW);
            return;
        }

        Files.write(path, data, StandardOpenOption.WRITE);
    }

    @Override
    public void delete() throws Throwable {
        Files.delete(path);
    }

    @Override
    public boolean exists() {
        return Files.exists(path, LinkOption.NOFOLLOW_LINKS);
    }

    @Override
    public MemoryFile getMemoryFile() {
        return defaultMemoryFile;
    }

    /**
     * Gets path.
     *
     * @return the path
     */
    public Path getPath() {
        return path;
    }
}

File: src/main/kotlin/net/cakemc/database/file/NioFile.kt
package net.cakemc.database.file

import java.io.IOException
import java.nio.file.Files
import java.nio.file.LinkOption
import java.nio.file.Path
import java.nio.file.StandardOpenOption

/**
 * The type Nio file.
 */
open class NioFile
/**
 * Instantiates a new Nio file.
 *
 * @param path       the path
 * @param compressed the compressed
 */(
    /**
     * The Path.
     */
    val path: Path,
    /**
     * The Compressed.
     */
    protected open val compressed: Boolean
) : AbstractDatabaseFile() {
    /**
     * Gets path.
     *
     * @return the path
     */

    /**
     * The Default memory file.
     */
    override var memoryFile: MemoryFile? = null
        protected set

    @Throws(IOException::class)
    override fun read() {
        this.memoryFile = DefaultMemoryFile(path, Files.readAllBytes(path), compressed)
    }

    @Throws(Throwable::class)
    override fun write() {
        val data = (if (this.memoryFile == null) DefaultMemoryFile(
            path, ByteArray(0),
            compressed
        ).also { memoryFile = it } else memoryFile)?.data

        if (!exists()) {
            if (data != null) {
                Files.write(path, data, StandardOpenOption.CREATE_NEW)
            }
            return
        }

        if (data != null) {
            Files.write(path, data, StandardOpenOption.WRITE)
        }
    }

    @Throws(Throwable::class)
    override fun delete() {
        Files.delete(path)
    }

    override fun exists(): Boolean {
        return Files.exists(path, LinkOption.NOFOLLOW_LINKS)
    }
}
